core/src/main/java/org/apache/brooklyn/core/typereg/AbstractTypePlanTransformer.java

```diff
@@ -0,0 +1,130 @@
+    public String getFormatDescription() {
+        return formatDescription;
+    }
+
+    @Override
+    public String toString() {
+        return getFormatCode()+":"+JavaClassNames.simpleClassName(this);
+    }
+    
+    @Override
+    public double scoreForType(RegisteredType type, RegisteredTypeLoadingContext context) {
+        if (getFormatCode().equals(type.getPlan().getPlanFormat())) return 1;
+        if (type.getPlan().getPlanFormat()==null)
+            return scoreForNullFormat(type.getPlan().getPlanData(), type, context);
+        else
+            return scoreForNonmatchingNonnullFormat(type.getPlan().getPlanFormat(), type.getPlan().getPlanData(), type, context);
+    }
+
+    protected abstract double scoreForNullFormat(Object planData, RegisteredType type, RegisteredTypeLoadingContext context);
+    protected abstract double scoreForNonmatchingNonnullFormat(String planFormat, Object planData, RegisteredType type, RegisteredTypeLoadingContext context);
+
+    /** delegates to more specific abstract create methods,
+     * and performs common validation and customisation of the items created.
+     * <p>
+     * this includes:
+     * <li> setting the {@link AbstractBrooklynObjectSpec#catalogItemId(String)}
+     */
+    @Override
+    public Object create(final RegisteredType type, final RegisteredTypeLoadingContext context) {
+        try {
+            return validate(new RegisteredTypeKindVisitor<Object>() {
+                @Override protected Object visitSpec(RegisteredType type) {
+                    try { 
+                        AbstractBrooklynObjectSpec<?, ?> result = createSpec(type, context);
+                        result.catalogItemId(type.getId());
+                        return result;
+                    } catch (Exception e) { throw Exceptions.propagate(e); }
+                }
+                @Override protected Object visitBean(RegisteredType type) {
+                    try { 
+                        return createBean(type, context);
+                    } catch (Exception e) { throw Exceptions.propagate(e); }
+                }
+                
+            }.visit(type), type, context);
+        } catch (UnsupportedTypePlanException e) {
+            // no logging
+            throw Exceptions.propagate(e);
+        } catch (Exception e) {
+            Exceptions.propagateIfFatal(e);
+            log.debug("Could not instantiate "+type+" (rethrowing): "+Exceptions.collapseText(e));
+            throw Exceptions.propagate(e);
+        }
+    }
+    
+    protected <T> T validate(T createdObject, RegisteredType type, RegisteredTypeLoadingContext context) {
+        if (createdObject==null) return null;
+        // TODO validation based on the constraint, throw UnsupportedTypePlanException with details if not matched
+        return createdObject;
+    }
+
+    protected abstract AbstractBrooklynObjectSpec<?,?> createSpec(RegisteredType type, RegisteredTypeLoadingContext context) throws Exception;
```

Comment:
there are so many exceptions when working with classes and loading, i don't think that's likely.  happy to revert in time however; i figure this will be beta for a few releases.